hal/inc/net_hal.h

```diff
@@ -47,8 +47,8 @@ typedef enum {
     NET_ACCESS_TECHNOLOGY_UTRAN = 4,
     NET_ACCESS_TECHNOLOGY_WCDMA = 4,
     NET_ACCESS_TECHNOLOGY_CDMA = 5,
-    NET_ACCESS_TECHNOLOGY_LTE = 6,
-    NET_ACCESS_TECHNOLOGY_IEEE802154 = 7,
+    NET_ACCESS_TECHNOLOGY_IEEE802154 = 6,
+    NET_ACCESS_TECHNOLOGY_LTE = 7,
```

Comment:
What's the reason for this change? It breaks the ABI.

Reply:
This unit test never passes when the RAT=7. 
https://github.com/particle-iot/device-os/blob/1f2d369d22d8af6f9ef25da3a3b0d59df5322ae9/user/tests/wiring/no_fixture/cellular.cpp#L155
Because it looks like it's getting mapped to `NET_ACCESS_TECHNOLOGY_IEEE802154` and never passes the unit test although the values are obtained.

Reply:
Do you know why? If that's the case then it has to be fixed either in the unit test or in the system code (if, for example, it exposes raw RAT identifiers obtained from the modem to the app code), but, again, we can't change values of an enum that is a part of a dynalib interface

Reply:
Thanks Sergey. Andrey walked me through some of the ABI stuff as well. 

I believe here's why:
The `s.getAccessTechnolog()` here (https://github.com/particle-iot/device-os/blob/6edbe275779e38cb68b8135f876ac1c1248d5758/user/tests/wiring/no_fixture/cellular.cpp#L122) gets the access technology from `sig_.rat` here (https://github.com/particle-iot/device-os/blob/6edbe275779e38cb68b8135f876ac1c1248d5758/wiring/src/spark_wiring_cellular_printable.cpp#L44) where `rat` here is `uint8_t` and is a directly obtained value from the cell radio - https://github.com/particle-iot/device-os/blob/6edbe275779e38cb68b8135f876ac1c1248d5758/hal/inc/cellular_hal_constants.h#L87
And we are directly converting the `uint8_t` to `hal_net_access_tech_t` for unit tests. 

This is not affecting the device-os `Cellular.RSSI()` itself because the `getStrengthValue` for example doesn't check for specific rat. But it affects the unit tests esp. only when RAT=7 on LTE radios (which is for LTE). There are two cases where RAT is `7` - always for LTE Cat-1 , and sometimes for LTE-M1 (ublox radios are strange).

I understand that changing the enum the way I did will break the ABI so I am going to revert this change, but I think this needs to be addressed for the unit tests to pass, if my understanding is correct. Will probably make a separate story/PR (probably having an intermediate translation between `getAccessTechnology` and `hal_net_access_tech_t`

Reply:
> where rat here is uint8_t and is a directly obtained value from the cell radio

Could you point me to where you think this might be happening in the code? Both [Gen 2](https://github.com/particle-iot/device-os/blob/develop/hal/src/electron/cellular_internal.cpp#L32-L54) and [Gen 3](https://github.com/particle-iot/device-os/blob/develop/hal/network/ncp/cellular/cellular_hal.cpp#L357) seem to be converting the modem's RAT identifiers to `hal_net_access_tech_t` rather than using them as is.

Reply:
The [Gen2](https://github.com/particle-iot/device-os/blob/develop/hal/src/electron/cellular_internal.cpp#L32-L54) converts into `hal_net_access_tech_t` but [here](https://github.com/particle-iot/device-os/blob/6edbe275779e38cb68b8135f876ac1c1248d5758/hal/src/electron/cellular_internal.cpp#L55) uses this [Act enum](https://github.com/particle-iot/device-os/blob/6edbe275779e38cb68b8135f876ac1c1248d5758/hal/shared/cellular_enums_hal.h#L82-L91) from `NetStatus` to make calculations. 
On [Gen3](https://github.com/particle-iot/device-os/blob/develop/hal/network/ncp/cellular/cellular_hal.cpp#L357), it uses `s.accessTechnolog()` which there gets values using `CellularAccessTechnolog` [enum](https://github.com/particle-iot/device-os/blob/6edbe275779e38cb68b8135f876ac1c1248d5758/hal/network/ncp/cellular/cellular_ncp_client.h#L69-L81). Although they are both converting RAT to `hal_net_access_tech_t`, they are using the other enums for calculations so we don't see any problem there. 

The problem seems to occur in unit tests which I put in [this](https://github.com/particle-iot/device-os/pull/2212/files#r504345008) comment. Hope that helps...

Reply:
Sorry, maybe I'm missing something, but I still don't understand how an incorrect RAT value (which the change that is being discussed in this thread was supposed to fix) gets returned to the application. Gen 2 uses `AcT` in calculations, that's correct, but it also sets `cellular_signal_t::rat` based on the same value of the `Act` enum. In other words, in case of LTE, the cellular HAL sets `cellular_signal_t::rat` to `NET_ACCESS_TECHNOLOGY_LTE` [here](https://github.com/particle-iot/device-os/blob/develop/hal/src/electron/cellular_internal.cpp#L43) and performs calculations specific to LTE [here](https://github.com/particle-iot/device-os/blob/develop/hal/src/electron/cellular_internal.cpp#L100-L114). Where's the error?

Reply:
The error is while running this unit test. Hope this provides more clarity.... 
While running [this](https://github.com/particle-iot/device-os/blob/6edbe275779e38cb68b8135f876ac1c1248d5758/user/tests/wiring/no_fixture/cellular.cpp#L112) unit test on a Boron LTE device when the RAT from the device is `7`:
- Gets the RAT [here](https://github.com/particle-iot/device-os/blob/6edbe275779e38cb68b8135f876ac1c1248d5758/user/tests/wiring/no_fixture/cellular.cpp#L122) from s.getAccessTechnoloy()
- This is [getAccessTechnolog()](https://github.com/particle-iot/device-os/blob/6edbe275779e38cb68b8135f876ac1c1248d5758/wiring/src/spark_wiring_cellular_printable.cpp#L44) - it reads the `sig_.rat`
- `rat` is `uint8_t` when stored is `7`, so `s.getAccessTechnolog()` gives `7`.
- Back to unit test - in the switch case [here](https://github.com/particle-iot/device-os/blob/6edbe275779e38cb68b8135f876ac1c1248d5758/user/tests/wiring/no_fixture/cellular.cpp#L122), the result of `s.getAccessTechnolog()` is `7`, it looks for `7` , doesn’t see it because `NET_ACCESS_TECHNOLOGY_LTE` is actually `6`

Reply:
> rat is uint8_t when stored is 7, so s.getAccessTechnolog() gives 7

That's the part I don't get. How 7 (`NET_ACCESS_TECHNOLOGY_IEEE802154`) ends up being stored in `rat` if we're not exposing raw RAT values reported by the modem. On Gen 3, we always convert raw values to `hal_net_access_tech_t` [here](https://github.com/particle-iot/device-os/blob/develop/hal/network/ncp/cellular/cellular_hal.cpp#L357) using [this function](https://github.com/particle-iot/device-os/blob/develop/hal/network/ncp/cellular/cellular_hal.cpp#L85-L106). As you can see that function never returns `NET_ACCESS_TECHNOLOGY_IEEE802154`.

Reply:
Thank you for helping check this, Sergey. Reverting the change back.